src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -138,37 +145,60 @@ public override void Initialize(AnalysisContext context)
             var enumeratedMethods = GetEnumeratedMethods(wellKnownTypeProvider, s_immutableCollectionsTypeNamesAndConvensionMethods, s_enumeratedParametersLinqMethods);
             var noEffectLinqChainMethods = GetLinqMethods(wellKnownTypeProvider, s_noEffectLinqChainMethods);
-            var compilation = operationBlockStartAnalysisContext.Compilation;
+            var compilation = compilationStartAnalysisContext.Compilation;
             var additionalDeferTypes = GetTypes(compilation, s_additionalDeferredTypes);
-
             // In CFG blocks there is no foreach loop related Operation, so use the
             // the GetEnumerator method to find the foreach loop
             var getEnumeratorSymbols = GetGetEnumeratorMethods(wellKnownTypeProvider);
+
+            var syntaxTree = compilation.SyntaxTrees.FirstOrDefault();
+            var customizedNoEnumerationMethods = syntaxTree == null
+                ? null
+                : compilationStartAnalysisContext.Options.GetNoEnumeratedMethodsOption(
+                    MultipleEnumerableDescriptor,
+                    syntaxTree,
+                    compilation);
+            var customizedLinqChainMethods = syntaxTree == null
+                ? null
+                : compilationStartAnalysisContext.Options.GetLinqChainMethodsOption(
+                    MultipleEnumerableDescriptor,
+                    syntaxTree,
+                    compilation);
+
             var wellKnownSymbolsInfo = new WellKnownSymbolsInfo(
-                deferredMethods,
+                linqChainMethods,
+                noEnumerationMethods,
                 enumeratedMethods,
                 noEffectLinqChainMethods,
                 additionalDeferTypes,
-                getEnumeratorSymbols);
+                getEnumeratorSymbols,
+                customizedNoEnumerationMethods,
+                customizedLinqChainMethods);
 
-            var potentialDiagnosticOperationsBuilder = PooledHashSet<IOperation>.GetInstance();
-            operationBlockStartAnalysisContext.RegisterOperationAction(
-                context => CollectPotentialDiagnosticOperations(
+            compilationStartAnalysisContext.RegisterOperationBlockStartAction(
+                context => OnOperationBlockStart(
                     context,
-                    wellKnownSymbolsInfo,
-                    potentialDiagnosticOperationsBuilder),
-                OperationKind.ParameterReference,
-                OperationKind.LocalReference);
+                    wellKnownTypeProvider,
+                    wellKnownSymbolsInfo));
+        }
 
+        private void OnOperationBlockStart(
+            OperationBlockStartAnalysisContext operationBlockStartAnalysisContext,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo)
+        {
+            var potentialDiagnosticOperationsBuilder = PooledHashSet<IOperation>.GetInstance();
```

Comment:
Instead of expecting this builder to be freed inside Analyze method at different return points, it might be more maintable and cleaner to wrap the `Analyze` invocation below within a try/finally, with `potentialDiagnosticOperationsBuilder.Free()` done in the finally.